hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java

```diff
@@ -595,10 +595,15 @@ private void onStartExecutionComplete(Throwable error, Collection<Entry<MemberIn
                     // have to use Async version, the future is completed inside a synchronized block
                     .whenCompleteAsync(withTryCatch(logger, (r, e) -> finalizeJob(finalError)));
         } else {
-            if (error instanceof ExecutionNotFoundException) {
-                // If the StartExecutionOperation didn't find the execution, it means that we must have cancelled it.
+            if (error instanceof ExecutionNotFoundException && requestedTerminationMode != null) {
+                // If the StartExecutionOperation didn't find the execution and requestedTerminationMode
+                // is not null it means that we must have cancelled it.
                 // Let's pretend that the StartExecutionOperation returned JobTerminateRequestedException
-                assert requestedTerminationMode != null && !requestedTerminationMode.isWithTerminalSnapshot();
+                // Note that the local execution cancellation need not be performed only on
+                // an explicit user cancellation request. In this case, requestedTerminationMode
+                // becomes null. For example, this local cancellation can happen in
+                // JobExecutionService#onMemberRemoved. We will continue to handle these cases as if we
+                // would handle an unspecific exception
                 error = new JobTerminateRequestedException(requestedTerminationMode);
```

Comment:
We must actually use something else because the `requestedTerminationMode` is null:
```suggestion
                TerminationMode mode = this.requestedTerminationMode;
                error = new JobTerminateRequestedException(mode != null ? mode : CANCEL_FORCEFUL);
```

Reply:
I changed the condition of the if block above. In this null case, I was throwing the `ExecutionNotFoundException` without changing it rather than a `JobTerminateRequestedException` so that a execution failure would occur. Maybe I can add an else block and turn this `ExecutionNotFoundException`  into a more understandable exception. What do you think if I convert it to `TopologyChangedException`?

Reply:
Ah, sorry, I missed that. Translating exceptions to other exceptions in general is dangerous - it often only confuses the failure analysis.

Now I think what you did is maybe the best option. If the exception was not found and we requested a cancellation, then we can convert. If we didn't request the cancellation, maybe it's better to keep the original exception.

Maybe it would be even good to put the ENFE as a cause of the JTRE.